subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1739 @@
+			goto end;
+		}
+		gc_count++;
+	}
+	rc = len;
+end:
+	k_mutex_unlock(&fs->zms_lock);
+	return rc;
+}
+
+int zms_delete(struct zms_fs *fs, uint32_t id)
+{
+	return zms_write(fs, id, NULL, 0);
+}
+
+ssize_t zms_read_hist(struct zms_fs *fs, uint32_t id, void *data, size_t len, uint32_t cnt)
+{
+	int rc, prev_found;
+	uint64_t wlk_addr, rd_addr;
+	uint32_t cnt_his;
+	struct zms_ate wlk_ate;
+	size_t ate_size;
+#ifdef CONFIG_ZMS_DATA_CRC
+	uint32_t computed_data_crc;
+#endif
+
+	if (!fs->ready) {
+		LOG_ERR("zms not initialized");
+		return -EACCES;
+	}
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	cnt_his = 0U;
+
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+	wlk_addr = fs->lookup_cache[zms_lookup_cache_pos(id)];
+
+	if (wlk_addr == ZMS_LOOKUP_CACHE_NO_ADDR) {
+		rc = -ENOENT;
+		goto err;
+	}
+#else
+	wlk_addr = fs->ate_wra;
+#endif
+	rd_addr = wlk_addr;
+
+	while (cnt_his <= cnt) {
+		rd_addr = wlk_addr;
+		/* Search for a previous valid ATE with the same ID */
+		prev_found =
+			zms_find_ate_with_id(fs, id, wlk_addr, fs->ate_wra, &wlk_ate, &rd_addr);
+		if (prev_found < 0) {
+			return prev_found;
+		}
+		if (prev_found) {
+			cnt_his++;
+			/* rd_addr contain the ATE address of the previous found ATE.
+			 * compute the previous ATE address in case we need to start
+			 * the research again.
+			 */
+			rc = zms_compute_prev_addr(fs, &rd_addr);
+			if (rc) {
+				return rc;
+			}
+			/* wlk_addr will be the start research address in the next loop */
+			wlk_addr = rd_addr;
+		} else {
+			wlk_addr = fs->ate_wra;
+			break;
+		}
+	}
+
+	if (((!prev_found) || (wlk_ate.id != id)) || (wlk_ate.len == 0U) || (cnt_his < cnt)) {
+		return -ENOENT;
+	}
+
+	if (wlk_ate.len <= ZMS_DATA_IN_ATE_SIZE) {
+		/* data is stored in the ATE */
+		memcpy(data, &wlk_ate.data, MIN(len, wlk_ate.len));
+	} else {
+		rd_addr &= ADDR_SECT_MASK;
+		rd_addr += wlk_ate.offset;
+		rc = zms_flash_rd(fs, rd_addr, data, MIN(len, wlk_ate.len));
+		if (rc) {
+			goto err;
+		}
+#ifdef CONFIG_ZMS_DATA_CRC
+		computed_data_crc = crc32_ieee(data, wlk_ate.len);
+		if (computed_data_crc != wlk_ate.data_crc) {
+			LOG_ERR("Invalid data CRC: ATE_CRC=0x%08X, computed_data_crc=0x%08X",
+				wlk_ate.data_crc, computed_data_crc);
+			return -EIO;
+		}
+#endif
+	}
+
+	return MIN(len, wlk_ate.len);
+
+err:
+	return rc;
+}
+
+ssize_t zms_read(struct zms_fs *fs, uint32_t id, void *data, size_t len)
+{
+	int rc;
+
+	rc = zms_read_hist(fs, id, data, len, 0);
+	return rc;
+}
+
+ssize_t zms_calc_free_space(struct zms_fs *fs)
+{
+
+	int rc, previous_sector_num = ZMS_INVALID_SECTOR_NUM, prev_found = 0;
+	struct zms_ate step_ate, wlk_ate;
+	uint64_t step_addr, wlk_addr, step_prev_addr, wlk_prev_addr;
+	uint8_t current_cycle;
+	size_t ate_size, free_space;
+
+	if (!fs->ready) {
+		LOG_ERR("zms not initialized");
+		return -EACCES;
+	}
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	free_space = 0;
+	for (uint32_t i = 1; i < fs->sector_count; i++) {
```

Comment:
Maybe we can also avoid the loop and just do
```
free_space = (fs->sector_count - 1) * (fs->sector_size - 4 * ate_size);
```
?

Reply:
You are reading my mind, I was about to change this :)